File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = [0]
        def postorder(node):
            if not node:
                return -1
            
            left = postorder(node.left)
            right = postorder(node.right)

            res[0] = max(res[0], 2 + left + right)

            return 1 + max(left, right)
        
        postorder(root)
        return res[0]

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxDia= 0;
    public int diameterOfBinaryTree(TreeNode root) {
        int[] maxDia = new int[1];
        postorder(root, maxDia);
        return maxDia[0];
    }
    
    public int postorder(TreeNode node, int[] maxDia) {
        if(node == null) {
            return -1;
        }
        
        int left = postorder(node.left, maxDia);
        int right = postorder(node.right, maxDia);
        maxDia[0] = Math.max(maxDia[0], left + right + 2);
        
        return Math.max(left, right) + 1;
        
    }
}